hs/src/Reach/Connector/ALGO.hs

```diff
@@ -1938,34 +1938,46 @@ cm km = \case
     when recordNew $
       addNewTok $ DLA_Var dv
     sallocVarLet (DLVarLet (Just vc) dv) sm (ce de) km
-  DL_ArrayMap at ansv aa lv iv (DLBlock _ _ body ra) -> do
+  DL_ArrayMap at ansv as xs iv (DLBlock _ _ body ra) -> do
     let anssz = typeSizeOf $ argTypeOf $ DLA_Var ansv
-    let (_, xlen) = argArrTypeLen aa
+    let xlen = arraysLength as
     let rt = argTypeOf ra
     check_concat_len anssz
     salloc_ (textyv ansv) $ \store_ans load_ans -> do
       cbs ""
       store_ans
       cfor xlen $ \load_idx -> do
         load_ans
-        doArrayRef at aa True $ Right load_idx
-        sallocLet lv (return ()) $
-          store_let iv True load_idx $ do
-            cp (ca ra >> ctobs rt) body
+        let arraysLoop xs' as' = case (xs', as') of
+              ([], []) -> do
+                cp (ca ra >> ctobs rt) body
+              (x:xsTail, a:asTail) -> do
+                doArrayRef at a True $ Right load_idx
+                sallocLet x (return ()) $
+                  store_let iv True load_idx $
+                  arraysLoop xsTail asTail
+              (_, _) -> impossible "DL_ArrayMap: inconsistent array sizes"
+        arraysLoop xs as
```

Comment:
I would rewrite this to be like `foldlM xyz go $ zip xs as` where `xyz` is the `cp ... body` part and `go` is basically the double cons case